### ISSUE TITLE
Load MuLaN from last checkpoint

### ISSUE THREAD

**Issue by ukemamaster** [lucidrains/musiclm-pytorch#17]
@lucidrains Could you please pass an argument to the `MuLaNTrainer` class to enable/disable loading from last available checkpoint?

**Comment by ukemamaster** [comment]
Alternatively if i do this: 
 
```python
model = torch.load(MODEL_SAVE_PATH + '/'  + last_checkpoint)

# create new OrderedDict that does not contain `module.`
from collections import OrderedDict
new_state_dict = OrderedDict()
for k, v in model.items():
    name = k[7:] # remove `module.`
    new_state_dict[name] = v

mulan.load_state_dict(new_state_dict)

trainer = MuLaNTrainer(
    mulan = mulan,
    dataset = dataset,
    num_train_steps = 50001,
    batch_size = 4,
    save_model_every = 5000,
    force_clear_prev_results = False,
)

trainer.train()
```
It throws the followinfg error: 

```python
  File "/venv/lib/python3.8/site-packages/musiclm_pytorch/musiclm_pytorch.py", line 168, in forward
    sim = einsum('b h i d, b h j d -> b h i j', q, k) * self.scale
RuntimeError: CUDA out of memory. Tried to allocate 232.00 MiB (GPU 0; 23.65 GiB total capacity; 16.83 GiB already allocated; 102.56 MiB free; 16.88 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

```
Plus, i am not sure if the trainer will start counting the `num_train_steps` from last checkpoint? or from 0 and override the already saved checkpoint?

**Comment by ukemamaster** [comment]
Solved by setting `trainer.register_buffer('steps', torch.Tensor([checkpoint_number]))` before `trainer.train()
`

**Comment by lucidrains** [comment]
@ukemamaster oh sure, you mean like automatically loading the last saved model from the results folder?

for your second comment, it works for me

are you sure your path is correct?

```python
trainer.save('./test.pt')
trainer.load('./test.pt')
```

**Comment by LWprogramming** [comment]
In my fork I have been doing [something similar](https://github.com/LWprogramming/audiolm-pytorch/blob/main/audiolm_pytorch_demo_backup.ipynb) when I was testing out training SoundStream.

I'm guessing the relevant issue is that, although these trainers do periodically save *a* checkpoint by default in `trainer.train()`, these checkpoints are *not* the same as the ones you'd get when you run `trainer.save()`, and the latter is what `trainer.load()` expects, not the default-saved ones. I got around this by using a separate directory (in the notebook, currently called `actual_soundstream_results`) to store the results of `trainer.save()` and totally ignoring the auto-generated checkpoints.

**Comment by ukemamaster** [comment]
@lucidrains 
> @ukemamaster oh sure, you mean like automatically loading the last saved model from the results folder?

Yes. but controlled by an argument. 
> for your second comment, it works for me
> 
> are you sure your path is correct?
> 
> ```python
> trainer.save('./test.pt')
> trainer.load('./test.pt')
> ```
It may work for single-gpu trained checkpoints. Have you tried for a multi-gpu saved checkpoint? I think it has something to do with the fact that the state_dict keys are `model.module.layer_names` in case of multi-gpu trained model (instead of `model.layer_names` in case of single-gpu).

**Comment by lucidrains** [comment]
> In my fork of AudioLM I have been doing [something similar](https://github.com/LWprogramming/audiolm-pytorch/blob/main/audiolm_pytorch_demo_backup.ipynb) when I was testing out training SoundStream.
> 
> I'm guessing the relevant issue is that, although these trainers do periodically save _a_ checkpoint by default in `trainer.train()`, these checkpoints are _not_ the same as the ones you'd get when you run `trainer.save()`, and the latter is what `trainer.load()` expects, not the default-saved ones. I got around this by using a separate directory (in the notebook, currently called `actual_soundstream_results`) to store the results of `trainer.save()` and totally ignoring the auto-generated checkpoints.

oh yes! you are absolutely right. i wasn't even using the right save function

should be fixed in this repository, let me fix audiolm too

**Comment by lucidrains** [comment]
@LWprogramming @ukemamaster ok, both should be fixed now!

**Comment by lucidrains** [comment]
@ukemamaster i'll get to the load checkpoint functionality tomorrow morning

**Comment by ukemamaster** [comment]
@lucidrains Now it seems fine. And i think there is no need to pass argument to the trainer class. 

The following works as expected: 
```python

trainer = MuLaNTrainer(
    mulan = mulan,
    dataset = dataset,
    num_train_steps = 50000,
    batch_size = 4,
    save_model_every = 1000,
    force_clear_prev_results = False,
    results_folder = MODEL_SAVE_PATH
)

checkpoints = os.listdir(MODEL_SAVE_PATH)

if checkpoints:
   trainer.load(MODEL_SAVE_PATH + last_checkpoint)

   # to start counting further steps from the loaded checkpoint
   trainer.register_buffer('steps', torch.Tensor([last_checkpoint_number]))

   print(f'Loaded from last checkpoint: [{last_checkpoint}]')

trainer.train()
```

**Comment by lucidrains** [comment]
@ukemamaster oh yes, that could be an issue. may have fixed this in 0.0.26, if you'd like to try again

**Comment by ukemamaster** [comment]
@lucidrains fixed. thanks
